hazelcast/src/main/java/com/hazelcast/map/impl/operation/MergeBackupOperation.java

```diff
@@ -0,0 +1,119 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.operation;
+
+import com.hazelcast.core.EntryView;
+import com.hazelcast.map.impl.MapDataSerializerHook;
+import com.hazelcast.map.impl.MapEntries;
+import com.hazelcast.map.impl.record.Record;
+import com.hazelcast.map.impl.record.RecordInfo;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.spi.BackupOperation;
+import com.hazelcast.spi.PartitionAwareOperation;
+import com.hazelcast.spi.merge.SplitBrainMergePolicy;
+import com.hazelcast.util.Clock;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.hazelcast.map.impl.EntryViews.createSimpleEntryView;
+import static com.hazelcast.map.impl.record.Records.applyRecordInfo;
+
+/**
+ * Creates backups for merged {@link MapEntries} after split-brain healing with a {@link SplitBrainMergePolicy}.
+ *
+ * @since 3.10
+ */
+public class MergeBackupOperation extends MapOperation implements PartitionAwareOperation, BackupOperation {
+
+    private MapEntries entries;
+    private List<RecordInfo> recordInfos;
+    private boolean hasWanReplication;
+
+    public MergeBackupOperation(String name, MapEntries entries, List<RecordInfo> recordInfos, boolean hasWanReplication) {
+        super(name);
+        this.entries = entries;
+        this.recordInfos = recordInfos;
+        this.hasWanReplication = hasWanReplication;
+    }
+
+    public MergeBackupOperation() {
+    }
+
+    @Override
+    public void run() {
+        for (int i = 0; i < entries.size(); i++) {
+            Data dataKey = entries.getKey(i);
+            Data dataValue = entries.getValue(i);
+            if (dataValue == null) {
+                recordStore.removeBackup(dataKey);
+                if (hasWanReplication) {
+                    mapEventPublisher.publishWanReplicationRemoveBackup(name, dataKey, Clock.currentTimeMillis());
+                }
+            } else {
+                Record record = recordStore.putBackup(dataKey, dataValue);
+                applyRecordInfo(record, recordInfos.get(i));
+                if (hasWanReplication) {
+                    Data dataValueAsData = mapServiceContext.toData(dataValue);
+                    EntryView entryView = createSimpleEntryView(dataKey, dataValueAsData, record);
+                    mapEventPublisher.publishWanReplicationUpdateBackup(name, entryView);
+                }
+                if (recordStore.shouldEvict()) {
```

Comment:
i was meaning, this if can be put between lines 70 and 71 above, if record-store is evictable no need to add any entry and continue looping.

secondly, we need to put this ` evict(dataKey)` in `afterRun` of the operation.

With these, we can get rid of continuous `evict` calls in the edge of memory usage.

Reply:
Can you please give the whole block as example?
```java
    @Override
    public void run() {
        for (int i = 0; i < entries.size(); i++) {
            Data dataKey = entries.getKey(i);
            Data dataValue = entries.getValue(i);
            if (dataValue == null) {
                recordStore.removeBackup(dataKey);
                if (hasWanReplication) {
                    mapEventPublisher.publishWanReplicationRemoveBackup(name, dataKey, Clock.currentTimeMillis());
                }
            } else {
                if (recordStore.shouldEvict()) {
                    // what should this do here? we cannot break, since we also might have
                    // backup removes in the entries, so the rule "after first eviction we can stop"
                    // doesn't apply here
                }
                Record record = recordStore.putBackup(dataKey, dataValue);
                applyRecordInfo(record, recordInfos.get(i));
                if (hasWanReplication) {
                    Data dataValueAsData = mapServiceContext.toData(dataValue);
                    EntryView entryView = createSimpleEntryView(dataKey, dataValueAsData, record);
                    mapEventPublisher.publishWanReplicationUpdateBackup(name, entryView);
                }
            }
        }
    }

    @Override
    public void afterRun() throws Exception {
        for (int i = 0; i < entries.size(); i++) {
            Data dataKey = entries.getKey(i);
            if (!recordStore.shouldEvict()) {
                break;
            }
            evict(dataKey);
        }
        recordStore.disposeDeferredBlocks();
    }
```
I have no clear picture what you have in mind, sorry.

Reply:
Like this: 
```
 @Override
    public void run() {
        for (int i = 0; i < entries.size(); i++) {
            Data dataKey = entries.getKey(i);
            Data dataValue = entries.getValue(i);
            if (dataValue == null) {
                recordStore.removeBackup(dataKey);
                if (hasWanReplication) {
                    mapEventPublisher.publishWanReplicationRemoveBackup(name, dataKey, Clock.currentTimeMillis());
                }
            } else {
                if (!recordStore.shouldEvict()) {
                    Record record = recordStore.putBackup(dataKey, dataValue);
                    applyRecordInfo(record, recordInfos.get(i));
                    if (hasWanReplication) {
                        Data dataValueAsData = mapServiceContext.toData(dataValue);
                        EntryView entryView = createSimpleEntryView(dataKey, dataValueAsData, record);
                        mapEventPublisher.publishWanReplicationUpdateBackup(name, entryView);
                    }
                }
            }
        }
    }

    @Override
    public void afterRun() throws Exception {
        evict(null);
        recordStore.disposeDeferredBlocks();
    }
```

Reply:
Ah, I didn't know that (a single) `evict(null)` would work. I had in mind, that just a single entry is evicted, but that is `ICache`, not `IMap`, isn't it?

Reply:
Yeah, it makes sense.